### PR TITLE
Implemented dynamic permission request

### PR DESCRIPTION
**Pull request by crackerizer**

### CODE REVIEW

**Comment by ArenaSource** [comment]
It would be nice to see this pull request merged. I'm currently using crackerizer's fork, so far so good.

**Comment by donaldguy** [comment]
Any updates on this? @agrieve ?

File: src/android/ChromeUsb.java [hunk 1]
```diff
@@ -113,10 +125,37 @@ public boolean execute(String action, CordovaArgs args, final CallbackContext ca
                 releaseInterface(args, params, callbackContext);
                 return true;
             } else if ("controlTransfer".equals(action)) {
-                controlTransfer(args, params, callbackContext);
+                new Thread(new Runnable() {
```
Comment (agrieve) [inline src/android/ChromeUsb.java]: nit: should use cordova.getThreadPool().execute(new Runnable()...)

File: src/android/ChromeUsb.java [hunk 2]
```diff
@@ -167,35 +204,70 @@ private void openDevice(CordovaArgs args, JSONObject params,
                 }
             }
             if (usbDev != null) {
-                if (!mUsbManager.hasPermission(usbDev)) {
-                    // TODO: Implement dynamic permission request.
-                    throw new UsbError("Permission request not yet implemented");
+                if(mUsbReceiver == null) {
```
Comment (agrieve) [inline src/android/ChromeUsb.java]: what happens if mUsbReceiver is not null? The first time it's created it's bound to the callbackContext of the caller, so it seems like subsequent calls to openDevice might not have the correct callbackContext called?

File: src/android/ChromeUsb.java [hunk 3]
```diff
@@ -167,35 +204,70 @@ private void openDevice(CordovaArgs args, JSONObject params,
...
+                                    } else {
+                                        Log.d(TAG, "permission denied for device " + device);
+                                    }
+                                }
+                            }
+                        }
+                    };
+
+                    IntentFilter filter = new IntentFilter(ACTION_USB_PERMISSION);
+                    webView.getContext().registerReceiver(mUsbReceiver, filter);
```
Comment (agrieve) [inline src/android/ChromeUsb.java]: There should be a matching unregisterReceiver for this. Should it be unregistered after it calls callbackContext.success()?

At the very least, it should be unregistered within onDestroy, like is done here: https://github.com/apache/cordova-plugin-network-information/blob/master/src/android/NetworkManager.java#L144

File: src/android/ChromeUsb.java [hunk 4]
```diff
@@ -367,19 +462,90 @@ boolean claimInterface(int interfaceNumber) {
         boolean releaseInterface(int interfaceNumber) {
             return mConnection.releaseInterface(mDevice.getInterface(interfaceNumber));
         }
-        int controlTransfer(int requestType, int request, int value, int index, byte[] buffer) {
-            return mConnection.controlTransfer(requestType, request, value, index,
-                    buffer, buffer.length, 0);
+        private static String getResultString(byte[] in) {
+            String ret = "";
+
+            for(int n = 0; n < in.length; n++) {
```
Comment (agrieve) [inline src/android/ChromeUsb.java]: Not sure if this is a performance-sensitive loop. If not, just add a comment saying so and switch to using a StringBuilder. If so, might be worth doing something like http://stackoverflow.com/questions/9655181/how-to-convert-a-byte-array-to-a-hex-string-in-java (except adding in a space char between bytes)

**Comment by agrieve** [comment]
Wow, awesome contribution! Sorry I didn't notice this the first time around :(.

In order to push the merge button, I'll need to you accept Google's contributor agreement online: https://github.com/MobileChromeApps/mobile-chrome-apps/blob/master/CONTRIBUTING.md

**Comment by fcorneli** [comment]
Any news on merging this PR? I also need chrome.usb.interruptTransfer within one of my projects.

**Comment by crackerizer** [comment]
Hi agrieve,

I already read the agreement and I accepted it! Hopefully, people can benefit from the code.

**Comment by fcorneli** [comment]
Agrieve, could this then be merged so we can start working on it from the original repo?

File: src/android/ChromeUsb.java (hunk 1 above)
Comment (fcorneli) [inline src/android/ChromeUsb.java]: This has been patched by:
https://github.com/FOODit/cordova-plugin-chrome-apps-usb/commit/5a7a5d932ad96f2c95045e45a8e63a183f8c5d49

File: src/android/ChromeUsb.java (hunk 2 above)
Comment (fcorneli) [inline src/android/ChromeUsb.java]: Indeed. I stumbled on this issue. Second call fails.

Comment (fcorneli) [inline src/android/ChromeUsb.java]: Quick and dirty fix for this issue available at: https://github.com/e-Contract/cordova-plugin-chrome-apps-usb/commit/f23ca8fd9d68425c568487d735fbf18b5ed10abb

**Comment by agrieve** [comment]
Hmm, @crackerizer, did you enter your github username when filling out the CLA? I don't see your github ID as having accepted the CLA on our side.

**Comment by fcorneli** [comment]
Hi @agrieve. You can also pull the master from
https://github.com/e-Contract/cordova-plugin-chrome-apps-usb
I've "signed" the CLA.

Only thing I'm not sure of is the interruptTransfer. See patch at:
https://github.com/e-Contract/cordova-plugin-chrome-apps-usb/commit/2b74b7dd0b325611ab3d7b35f9a92af57f0650bb
I had to change this so interrupt IN works, but I guess the code that was in place was supposed to implement interrupt OUT. Maybe we need to "merge" both behaviors, depending on the direction.

**Comment by crackerizer** [comment]
Sorry, I have filled the CLA now.

**Comment by agrieve** [comment]
This is now merged!

**Pull request closed without merging**